### ISSUE TITLE
SaveSessionHandler saves the session too early when using a StreamedResponse

### ISSUE THREAD

**Issue by bestform** [symfony/symfony#12423]
In 9c1e4686 a new response listener was introduced by @Tobion , that saves the session on a kernel response event. When using a streamed response, this is actually pretty early, as most of the code is handled in the callback, that is being called after the response event was fired.

We ran into a problem with this setup as we used "isStarted" on the session instance, which of course is false after said listener has done its work.

It is argued in the commit message, that saving the session at an arbitrary point in time isn't critical, as it will be started again on demand. In the case of a StreamedResponse this leads to starting the session twice on every request. Moreover the return value of isStarted() on the session instance - while being technically correct - is quite misleading, as the session has in fact been started already and is just waiting to be restarted on demand after closing it via the save() call in the new listener.

I don't know a good solution for this from the top of my head but I would like to see it being discussed further, as the situation, that arises when using a StreamedResponse, isn't ideal at all.

**Comment by nikolajosipovic** [comment]
I'm doing a sub-request (->forward) within the StreamedResponse callback, so the Session is failing to start: `Failed to start session because headers already sent by..`

One solution was to override the introduced save session listener. Any other workaround? Problem is that the session data is required in the sub-request.

**Comment by nicolas-grekas** [comment]
I'm closing here because the issue is old, and the current behavior looks sensible to me: reading/writing to the session should be done before sending the first byte. Streamed responses need to be built around this IMHO. This also ensures that session is not locked for too long.

**Comment by AndrolGenhald** [comment]
@nicolas-grekas This makes sense to me, but the [documentation](https://symfony.com/doc/current/migration.html#legacy-route-loader) for doing a legacy route loader suggests using a StreamedResponse, which causes issues when using the session in the legacy code. Should this be reopened, or should that documentation be fixed with an alternative suggested implementation?

Edit: Perhaps it's a non-issue, I think the problem is that I removed the `session_start()` call in the legacy code when I did the Legacy Bridge, but it needed to be added back when migrating to the Legacy Route Loader.
